### PR TITLE
Implement bounding volume intersections

### PR DESCRIPTION
**Pull request by Jondolf**
# Objective

#10946 added bounding volume types and an `IntersectsVolume` trait, but didn't actually implement intersections between bounding volumes.

This PR implements AABB-AABB, circle-circle / sphere-sphere, and AABB-circle / AABB-sphere intersections.

## Solution

Implement `IntersectsVolume` for bounding volume pairs. I also added `closest_point` methods to return the closest point on the surface / inside of bounding volumes. This is used for AABB-circle / AABB-sphere intersections.

### CODE REVIEW

File: crates/bevy_math/src/bounding/bounded2d/mod.rs [hunk 1]
```diff
@@ -129,10 +139,32 @@ impl BoundingVolume for Aabb2d {
...
+    }
+}
+
+impl IntersectsVolume<BoundingCircle> for Aabb2d {
+    #[inline(always)]
+    fn intersects(&self, circle: &BoundingCircle) -> bool {
+        let closest_point = self.closest_point(circle.center);
+        let distance_squared = circle.center.distance_squared(closest_point);
+        let radius_squared = circle.radius().powi(2);
+        distance_squared < radius_squared
```
Comment (IQuick143) [inline crates/bevy_math/src/bounding/bounded2d/mod.rs:157]: This is nitpicking at edge cases, but I'd say this should be less than or equals, to account for the points at the surface of the ball.
```suggestion
        distance_squared <= radius_squared
```

File: crates/bevy_math/src/bounding/bounded3d/mod.rs [hunk 2]
```diff
@@ -125,10 +135,33 @@ impl BoundingVolume for Aabb3d {
...
+    }
+}
+
+impl IntersectsVolume<BoundingSphere> for Aabb3d {
+    #[inline(always)]
+    fn intersects(&self, sphere: &BoundingSphere) -> bool {
+        let closest_point = self.closest_point(sphere.center);
+        let distance_squared = sphere.center.distance_squared(closest_point);
+        let radius_squared = sphere.radius().powi(2);
+        distance_squared < radius_squared
```
Comment (IQuick143) [inline crates/bevy_math/src/bounding/bounded3d/mod.rs:154]: Same here.
```suggestion
        distance_squared <= radius_squared
```

**Review by IQuick143 — CHANGES_REQUESTED** [review]
For consistency I'd change the < to <=, but other than that I haven't noticed any problems.

**Review by NiseVoid — COMMENTED** [review]
This seems correct. Didn't check the tests tho. And `closest_point` on sphere/circle appears to be unused. It might be useful, but I also feel like maybe it should be on the primitive types instead.

**Comment by Jondolf** [comment]
For now, I moved the circle/sphere `closest_points` logic to the primitives (same for AABBs), but kept the methods on the bounding volumes for consistency and convenience. I can also just remove the circle/sphere version since it's unused, or split it into a different PR, but I thought it'd be nice to have if AABBs have it anyway

**Review by IQuick143 — APPROVED** [review]
Seems good.

Although I'm not super convinced about implementing closest point on the primitives, because closest point seems to demand transform information (translation + rotation), which is addressed here by assuming default identity transform.
It's not a huge deal, because the user can use this to implement the general case simply by inversely transforming the point and then transforming the result.

**Comment by Jondolf** [comment]
FWIW, Parry has "global" and local versions of a ton of methods, like `aabb`/`local_aabb`, `project_point`/`project_local_point`, `support_point`/`local_support_point`, and so on. Both can be useful.

The `closest_point` methods on the primitives currently correspond the most to `project_local_point`, but with the shape always treated as "solid" (the returned point can be inside the shape instead of always being at the perimeter/surface). In the future we might want a trait for this and to implement it for all primitives

**Comment by Jondolf** [comment]
Here's a video of the bounding volume intersections working: (gray is the bounding volume)

https://github.com/bevyengine/bevy/assets/57632562/08bfdb32-8408-43ea-a13e-c9fb650e4b54

I can add this as an example if that'd be useful. Should I do it in this PR or a follow-up?

**Comment by NiseVoid** [comment]
> Should I do it in this PR or a follow-up?

I guess a follow up PR that showcases the intersection tests (including raycasting and later shapecasting) and bounding types makes sense?

**Review by alice-i-cecile — APPROVED** [review]
Looks good. Great docs, and thanks for including thorough tests.

**Pull request merged**